File: core/r/R/heatmap.R
# Recipe: clustered heatmap of a features-by-samples matrix (ComplexHeatmap).
# Input is a matrix CSV: first column = feature id, remaining columns = samples.
# An optional annotation CSV (--annotation) labels the columns; its first column
# is the sample id. By default rows are z-scored and both axes are clustered.
# When the primary annotation column has exactly two groups, each feature gets a
# Welch t-test (BH-adjusted) and significance stars are drawn beside it.

recipe_heatmap <- function(df, spec) {
  feat <- as.character(df[[1]])
  mat  <- as.matrix(df[, -1, drop = FALSE])
  storage.mode(mat) <- "double"
  rownames(mat) <- feat

  n0 <- nrow(mat)
  keep <- apply(mat, 1, function(r) all(is.finite(r)) && stats::sd(r) > 0)
  mat <- mat[keep, , drop = FALSE]
  clean_steps <- if (sum(!keep) > 0) {
    sprintf("Dropped %d feature(s) with missing values or zero variance (%d -> %d).",
            sum(!keep), n0, nrow(mat))
  } else "No cleaning applied; input used as-is."
  df_used <- df[df[[1]] %in% rownames(mat), , drop = FALSE]

  scale_mode <- spec$appearance$scale %||% "row"
  cluster    <- spec$appearance$cluster %||% "both"
  disp <- switch(scale_mode,
    row    = t(scale(t(mat))),
    column = scale(mat),
    none   = mat,
    t(scale(t(mat))))
  disp[!is.finite(disp)] <- 0
  legend_name <- if (scale_mode == "none") "value" else "z-score"

  cluster_rows <- cluster %in% c("both", "rows")
  cluster_cols <- cluster %in% c("both", "columns")

  # Optional column annotation.
  top_anno <- NULL; ann <- NULL; group_col <- NULL
  if (!is.null(spec$data$annotation) && nzchar(spec$data$annotation)) {
    araw <- read_tidy(spec$data$annotation)
    rn <- as.character(araw[[1]])
    ann <- araw[, -1, drop = FALSE]
    rownames(ann) <- rn
    ann <- ann[colnames(mat), , drop = FALSE]
    for (cn in names(ann)) ann[[cn]] <- as.character(ann[[cn]])
    ann_cols <- lapply(names(ann), function(cn) {
      lv <- sort(unique(ann[[cn]]))
      stats::setNames(rep(pubplot_palette, length.out = length(lv)), lv)
    })
    names(ann_cols) <- names(ann)
    top_anno <- ComplexHeatmap::HeatmapAnnotation(
      df = ann, col = ann_cols,
      annotation_name_gp = grid::gpar(fontsize = 9))
    group_col <- names(ann)[1]
  }

  # Per-feature differential test when the primary annotation has two groups.
  right_anno <- NULL
  if (!is.null(group_col) && length(unique(ann[[group_col]])) == 2) {
    g  <- factor(ann[[group_col]]); lv <- levels(g)
    p  <- apply(mat, 1, function(r)
      tryCatch(stats::t.test(r[g == lv[1]], r[g == lv[2]])$p.value,
               error = function(e) NA_real_))
    padj  <- stats::p.adjust(p, "BH")
    stars <- ifelse(is.na(padj), "",
             ifelse(padj < 0.001, "***", ifelse(padj < 0.01, "**",
             ifelse(padj < 0.05, "*", ""))))
    right_anno <- ComplexHeatmap::rowAnnotation(
      ` ` = ComplexHeatmap::anno_text(stars, gp = grid::gpar(fontsize = 10)))
    stats_df <- data.frame(
      feature = rownames(mat),
      mean_g1 = rowMeans(mat[, g == lv[1], drop = FALSE]),
      mean_g2 = rowMeans(mat[, g == lv[2], drop = FALSE]),
      p_value = p, p_adj = padj, significant = padj < 0.05,
      stringsAsFactors = FALSE)
    names(stats_df)[2:3] <- paste0("mean_", lv)
    test_label <- sprintf("per-feature Welch t-test (%s vs %s), BH-adjusted", lv[1], lv[2])
    n_sig <- sum(padj < 0.05, na.rm = TRUE)
  } else {
    stats_df <- data.frame(feature = rownames(mat), mean = rowMeans(mat),
                           sd = apply(mat, 1, stats::sd), stringsAsFactors = FALSE)
    test_label <- NULL; n_sig <- NA_integer_
  }

  col_fun <- if (scale_mode == "none") {
    q <- stats::quantile(disp, c(0.02, 0.5, 0.98), na.rm = TRUE)
    circlize::colorRamp2(q, c("#3B6DB3", "white", "#C1432B"))
  } else {
    circlize::colorRamp2(c(-2, 0, 2), c("#3B6DB3", "white", "#C1432B"))
  }

  ht <- ComplexHeatmap::Heatmap(
    disp, name = legend_name, col = col_fun,
    cluster_rows = cluster_rows, cluster_columns = cluster_cols,
    top_annotation = top_anno, right_annotation = right_anno,
    row_names_gp = grid::gpar(fontsize = 8),
    column_names_gp = grid::gpar(fontsize = 8),
    heatmap_legend_param = list(title_gp = grid::gpar(fontsize = 9, fontface = "bold")))
  draw_fn <- function() ComplexHeatmap::draw(ht, merge_legends = TRUE)

  test_meta <- list(
    name = "Clustered heatmap (ComplexHeatmap)",
    scale = scale_mode, clustering = cluster,
    n_features = nrow(mat), n_samples = ncol(mat),
    per_feature_test = test_label, n_significant = n_sig)

  resolved <- list(family = "heatmap", label = "heatmap")
  methods  <- .hm_methods(scale_mode, cluster, nrow(mat), ncol(mat), test_label, n_sig)
  build_script <- function(in_name, fig_stub) {
    .hm_emit_script(spec, in_name, fig_stub, scale_mode, cluster_rows, cluster_cols, legend_name)
  }

  list(plot = NULL, draw = draw_fn, stats = stats_df, test_meta = test_meta,
       resolved = resolved, methods = methods, build_script = build_script,
       label = slugify(tools::file_path_sans_ext(spec$data$file)),
       df_used = df_used, clean_steps = clean_steps,
       extra_inputs = spec$data$annotation,
       width = max(5.5, 2.2 + 0.24 * ncol(mat)),
       height = max(5.0, 1.5 + 0.16 * nrow(mat)))
}

.hm_methods <- function(scale_mode, cluster, n_feat, n_samp, test_label, n_sig) {
  v    <- pkg_versions(c("ComplexHeatmap", "circlize"))
  rver <- paste(R.version$major, R.version$minor, sep = ".")
  scale_txt <- switch(scale_mode,
    row = "Values were z-scored across samples within each feature (row scaling).",
    column = "Values were z-scored within each sample (column scaling).",
    "Raw values were shown without scaling.")
  clust_txt <- switch(cluster,
    both = "Rows and columns were hierarchically clustered.",
    rows = "Rows were hierarchically clustered; column order was preserved.",
    columns = "Columns were hierarchically clustered; row order was preserved.",
    "No clustering was applied.")
  test_txt <- if (!is.null(test_label)) {
    sprintf(" A %s was applied per feature; %d feature(s) were significant (adjusted P < 0.05) and are marked with stars.",
            test_label, n_sig)
  } else ""
  paste0(
    sprintf("A heatmap of %d features across %d samples was drawn with ComplexHeatmap. ",
            n_feat, n_samp),
    scale_txt, " ", clust_txt, test_txt,
    sprintf(" Rendered in R %s with ComplexHeatmap %s and circlize %s.",
            rver, v$ComplexHeatmap, v$circlize))
}

.hm_emit_script <- function(spec, in_name, fig_stub, scale_mode, cluster_rows, cluster_cols, legend_name) {
  ann_name <- if (!is.null(spec$data$annotation) && nzchar(spec$data$annotation)) {
    basename(spec$data$annotation)
  } else NULL
  disp_line <- switch(scale_mode,
    row    = "disp <- t(scale(t(mat)))",
    column = "disp <- scale(mat)",
    "disp <- mat")
  ann_lines <- if (!is.null(ann_name)) c(
    sprintf('ann <- read.csv("%s", check.names = FALSE, row.names = 1)', ann_name),
    "ann <- ann[colnames(mat), , drop = FALSE]",
    "top <- HeatmapAnnotation(df = ann)"
  ) else "top <- NULL"

  c(
    "#!/usr/bin/env Rscript",
    "# Standalone reproduction. Run inside the pinned pubplot container",
    "# (see REPRODUCE.md) from this bundle folder.",
    "suppressPackageStartupMessages({ library(ComplexHeatmap); library(circlize) })",
    "",
    sprintf('raw <- read.csv("%s", check.names = FALSE, row.names = 1)', in_name),
    "mat <- as.matrix(raw); storage.mode(mat) <- \"double\"",
    "keep <- apply(mat, 1, function(r) all(is.finite(r)) && sd(r) > 0); mat <- mat[keep, ]",
    disp_line,
    "disp[!is.finite(disp)] <- 0",
    ann_lines,
    sprintf('col_fun <- colorRamp2(c(-2, 0, 2), c("#3B6DB3", "white", "#C1432B"))'),
    sprintf('ht <- Heatmap(disp, name = "%s", col = col_fun, cluster_rows = %s, cluster_columns = %s, top_annotation = top)',
            legend_name, if (cluster_rows) "TRUE" else "FALSE", if (cluster_cols) "TRUE" else "FALSE"),
    "",
    sprintf('cairo_pdf("%s.pdf", width = 7, height = 7)', fig_stub),
    "draw(ht, merge_legends = TRUE); dev.off()",
    'cat("Reproduced figure.\\n")'
  )
}

File: core/python/pubready/recipes/heatmap.py
"""Recipe: clustered heatmap of a features-by-samples matrix. Tries
PyComplexHeatmap (to mirror the R ComplexHeatmap output); falls back to a seaborn
clustermap if PyComplexHeatmap raises. Row z-score and clustering by default;
a per-feature Welch t-test (BH) is computed when the annotation has two groups."""
import os

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats
from statsmodels.stats.multitest import multipletests

from ..bundle import slugify
from ..io import read_tidy
from ..theme import PALETTE


def recipe_heatmap(df, spec):
    feat = df.iloc[:, 0].astype(str)
    mat = df.iloc[:, 1:].apply(pd.to_numeric, errors="coerce")
    mat.index = list(feat)
    n0 = len(mat)
    keep = mat.notna().all(axis=1) & (mat.std(axis=1) > 0)
    mat = mat[keep]
    clean_steps = (f"Dropped {n0 - len(mat)} feature(s) with missing values or zero variance "
                   f"({n0} -> {len(mat)})." if len(mat) < n0 else "No cleaning applied; input used as-is.")
    df_used = df[df.iloc[:, 0].astype(str).isin(mat.index)].reset_index(drop=True)

    ap = spec["appearance"]
    scale_mode = ap.get("scale") or "row"
    cluster = ap.get("cluster") or "both"
    if scale_mode == "row":
        disp = mat.sub(mat.mean(axis=1), axis=0).div(mat.std(axis=1), axis=0)
    elif scale_mode == "column":
        disp = (mat - mat.mean()) / mat.std()
    else:
        disp = mat.copy()
    disp = disp.fillna(0.0)
    legend_name = "value" if scale_mode == "none" else "z-score"
    cluster_rows = cluster in ("both", "rows")
    cluster_cols = cluster in ("both", "columns")

    ann, group_col = None, None
    if spec["data"].get("annotation"):
        araw = read_tidy(spec["data"]["annotation"])
        araw = araw.set_index(araw.columns[0])
        ann = araw.reindex(mat.columns)
        ann = ann.astype(str)
        group_col = ann.columns[0]

    # Per-feature differential test when the primary annotation has two groups.
    if ann is not None and ann[group_col].nunique() == 2:
        g = ann[group_col]
        lv = sorted(g.unique())
        ca = [c for c in mat.columns if g.loc[c] == lv[0]]
        cb = [c for c in mat.columns if g.loc[c] == lv[1]]
        pvals = mat.apply(lambda r: stats.ttest_ind(r[ca], r[cb], equal_var=False).pvalue, axis=1).to_numpy()
        padj = multipletests(pvals, method="fdr_bh")[1]
        stats_df = pd.DataFrame({
            "feature": mat.index, f"mean_{lv[0]}": mat[ca].mean(axis=1).to_numpy(),
            f"mean_{lv[1]}": mat[cb].mean(axis=1).to_numpy(),
            "p_value": pvals, "p_adj": padj, "significant": padj < 0.05,
        })
        test_label = f"per-feature Welch t-test ({lv[0]} vs {lv[1]}), BH-adjusted"
        n_sig = int((padj < 0.05).sum())
    else:
        stats_df = pd.DataFrame({"feature": mat.index, "mean": mat.mean(axis=1).to_numpy(),
                                 "sd": mat.std(axis=1).to_numpy()})
        test_label, n_sig = None, None

    w = max(5.5, 2.2 + 0.24 * mat.shape[1])
    h = max(5.0, 1.5 + 0.16 * mat.shape[0])

    used_pch = True
    try:
        import PyComplexHeatmap as pch
        col_ha = None
        if ann is not None:
            colors = {lvl: PALETTE[i % len(PALETTE)]
                      for i, lvl in enumerate(sorted(ann[group_col].unique()))}
            col_ha = pch.HeatmapAnnotation(
                **{group_col: pch.anno_simple(ann[group_col], colors=colors, add_text=False)},
                axis=1, verbose=0)
        plt.figure(figsize=(w, h))
        pch.ClusterMapPlotter(
            data=disp, top_annotation=col_ha,
            row_cluster=cluster_rows, col_cluster=cluster_cols,
            cmap="RdBu_r", label=legend_name,
            show_rownames=True, show_colnames=True,
            row_names_side="right", col_names_side="bottom", verbose=0)
        fig = plt.gcf()
    except Exception:
        plt.close("all")
        used_pch = False
        col_colors = None
        if ann is not None:
            colors = {lvl: PALETTE[i % len(PALETTE)]
                      for i, lvl in enumerate(sorted(ann[group_col].unique()))}
            col_colors = ann[group_col].map(colors)
        grid = sns.clustermap(
            disp, cmap="RdBu_r", center=0, row_cluster=cluster_rows, col_cluster=cluster_cols,
            col_colors=col_colors, figsize=(w, h), cbar_kws={"label": legend_name},
            xticklabels=True, yticklabels=True)
        fig = grid.figure

    test_meta = {
        "name": "Clustered heatmap", "engine_lib": "PyComplexHeatmap" if used_pch else "seaborn.clustermap",
        "scale": scale_mode, "clustering": cluster,
        "n_features": int(mat.shape[0]), "n_samples": int(mat.shape[1]),
        "per_feature_test": test_label, "n_significant": n_sig,
    }
    methods = _methods(scale_mode, cluster, mat.shape[0], mat.shape[1], test_label, n_sig, used_pch)
    return {
        "fig": fig, "stats": stats_df, "test_meta": test_meta, "methods": methods,
        "df_used": df_used, "clean_steps": clean_steps,
        "label": slugify(os.path.splitext(spec["data"]["file"])[0]),
        "extra_inputs": [spec["data"].get("annotation")],
        "width": w, "height": h,
        "build_script": lambda in_name, fig_stub: _script(spec, scale_mode, cluster_rows, cluster_cols, legend_name, in_name, fig_stub),
    }


def _methods(scale_mode, cluster, n_feat, n_samp, test_label, n_sig, used_pch):
    from importlib.metadata import version as v
    import platform
    lib = "PyComplexHeatmap" if used_pch else "seaborn.clustermap"
    lib_ver = v("PyComplexHeatmap") if used_pch else v("seaborn")
    scale_txt = {"row": "Values were z-scored across samples within each feature (row scaling).",
                 "column": "Values were z-scored within each sample (column scaling).",
                 "none": "Raw values were shown without scaling."}.get(scale_mode)
    clust_txt = {"both": "Rows and columns were hierarchically clustered.",
                 "rows": "Rows were hierarchically clustered.",
                 "columns": "Columns were hierarchically clustered.",
                 "none": "No clustering was applied."}.get(cluster)
    test_txt = (f" A {test_label} was applied per feature; {n_sig} feature(s) were significant "
                "(adjusted P < 0.05)." if test_label else "")
    return (
        f"A heatmap of {n_feat} features across {n_samp} samples was drawn with {lib}. "
        + scale_txt + " " + clust_txt + test_txt +
        f" Rendered in Python {platform.python_version()} with {lib} {lib_ver}."
    )


def _script(spec, scale_mode, cluster_rows, cluster_cols, legend_name, in_name, fig_stub):
    ann = spec["data"].get("annotation")
    ann_name = os.path.basename(ann) if ann else None
    disp_line = {"row": "disp = mat.sub(mat.mean(axis=1), axis=0).div(mat.std(axis=1), axis=0).fillna(0)",
                 "column": "disp = ((mat - mat.mean()) / mat.std()).fillna(0)",
                 "none": "disp = mat"}.get(scale_mode)
    ann_lines = (f'ann = pd.read_csv("{ann_name}", index_col=0).astype(str).reindex(mat.columns)\n'
                 f'gc = ann.columns[0]\n'
                 f'colors = {{lvl: c for lvl, c in zip(sorted(ann[gc].unique()), ["#3B6DB3", "#C1432B", "#2E8B57"])}}\n'
                 f'col_colors = ann[gc].map(colors)') if ann_name else "col_colors = None"
    return f'''#!/usr/bin/env python3
# Standalone reproduction (seaborn clustermap). Run inside the pinned container.
import matplotlib; matplotlib.use("Agg")
import pandas as pd, seaborn as sns

raw = pd.read_csv("{in_name}", index_col=0)
mat = raw.apply(pd.to_numeric, errors="coerce")
mat = mat[mat.notna().all(axis=1) & (mat.std(axis=1) > 0)]
{disp_line}
{ann_lines}
g = sns.clustermap(disp, cmap="RdBu_r", center=0, row_cluster={cluster_rows}, col_cluster={cluster_cols},
                   col_colors=col_colors, cbar_kws={{"label": "{legend_name}"}},
                   xticklabels=True, yticklabels=True)
g.figure.savefig("{fig_stub}.pdf")
print("Reproduced figure.")
'''
